sg/SG.3
.\"
.\" Copyright (c) 2006-2022 Julien Nadeau Carriere <vedge@csoft.net>
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 
.\" THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
.\" IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
.\" WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT,
.\" INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
.\" (INCLUDING BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
.\" SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
.\" STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING
.\" IN ANY WAY OUT OF THE USE OF THIS SOFTWARE EVEN IF ADVISED OF THE
.\" POSSIBILITY OF SUCH DAMAGE.
.\"
.Dd December 21, 2022
.Dt SG 3
.Os Agar 1.7
.Sh NAME
.Nm SG
.Nd Agar-SG 3D scene graph
.Sh SYNOPSIS
.Bd -literal
#include <agar/core.h>
#include <agar/sg.h>
.Ed
.Sh DESCRIPTION
The
.Nm
object is a general-purpose scene-graph, or a collection of nodes (i.e.,
.Xr SG_Node 3
objects) organized in a tree structure.
The
.Ft SG_Node
object is used to describe both "group" nodes and "leaf" nodes.
.Pp
Scene-graphs are usually rendered using the
.Xr SG_View 3
widget (or a subclass of it).
.Sh INHERITANCE HIERARCHY
.Xr AG_Object 3 ->
.Nm .
.Sh INITIALIZATION
.nr nS 1
.Ft "SG *"
.Fn SG_New "void *parent" "const char *name" "Uint flags"
.Pp
.Ft "void"
.Fn SG_Clear "SG *sg"
.Pp
.nr nS 0
The
.Fn SG_New
function allocates, initializes, and attaches a
.Nm
object.
Acceptable
.Fa flags
options include:
.Bl -tag -width "SG_SKIP_UNKNOWN_NODES "
.It SG_OVERLAY_WIREFRAME
Overlay wireframe onto geometric objects when rendering scene.
.It SG_OVERLAY_VERTICES
Overlay vertices onto geometric objects when rendering scene.
.It SG_OVERLAY_VNORMALS
Overlay vertex normals when rendering scene.
.It SG_OVERLAY_FNORMALS
Overlay facet normals when rendering scene.
.It SG_SKIP_UNKNOWN_NODES
When loading the contents of a scene, ignore unknown/unimplemented
elements (otherwise, encountering an unknown node will make the load
operation fail).
.It SG_NO_DEFAULT_NODES
Don't create the default nodes (e.g.,
.Sq Camera0 ,
.Sq Light0 ) .
.El
.Pp
The
.Fn SG_Clear
routine reinitializes
.Fa sg
to an empty scene.
If default nodes (e.g.,
.Sq Camera0 )
are used, they are preserved but their state is reinitialized.
.Sh NODE QUERIES
.nr nS 1
.Ft "SG_Node *"
.Fn SG_SearchNodes "SG_Node *parent" "const char *name"
.Pp
.Ft "void *"
.Fn SG_FindNode "SG *sg" "const char *name"
.Pp
.Ft "int"
.Fn SG_Intersect "SG_Node *node" "M_Geom3 g" "M_GeomSet3 *S"
.Pp
.nr nS 0
The
.Fn SG_SearchNodes
function scans all descendants of
.Fa parent
for a node of the specified name.
The
.Fn SG_FindNode
function searches the entire scene graph
.Fa sg
for a node of the specified name.
On success,
.Fn SG_SearchNodes
and
.Fn SG_FindNode
return a pointer to the matching node.
.Pp
The
.Fn SG_Intersect
function performs an intersection test between the given node and an
arbitrary geometrical object
.Fa g
(see
.Xr M_Geom3 3 ) .
The function returns 1 if the objects intersect, 0 if they don't
intersect, or -1 if the test is not implemented.
If the
.Fa S
argument is non-NULL, the intersections are computed and returned as one
or more geometrical objects, into
.Fa S .
The structure pointed to by
.Fa S
should have been previously initialized with
.Xr M_GEOM_SET_EMPTY 3 .
.Sh SEE ALSO
.Xr sgedit 1 ,
.Xr M_Matrix 3 ,
.Xr M_Plane 3 ,
.Xr M_Real 3 ,
.Xr M_Vector 3 ,
.Xr SG_Intro 3
.Sh HISTORY
The
.Nm
library first appeared in Agar 1.6.0.


CHANGELOG.md
# Changelog
All notable changes to Agar will be documented in this file. The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0), and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]
- The Agar Interface Builder [**agarib**](https://libagar.org/man1/agarib).
- The Agar Debugger [**agardb**](https://libagar.org/man1/agardb).
- Ada: Add Widget classes to the new Ada bindings in `ada/gui/`.
- New COBOL bindings: [GnuCOBOL](https://sourceforge.net/projects/open-cobol/) and [GuiCOBOL](http://www.opencobol.altervista.org).
- [**AG_Console**](https://libagar.org/man3/AG_Console): Handle ansi cursor and terminal operations. Clipboard integration.
- [**AG_FontSelector**](https://libagar.org/man3/AG_FontSelector): Display only those styles that are available for a given font (as opposed to a static list of styles).
- [**AG_Notebook**](https://libagar.org/man3/AG_Notebook): Fix padding issues. Add disposition modes Bottom, Left & Right.
- [**AG_Pixmap**](https://libagar.org/man3/AG_Pixmap) & [**AG_Fixed**](https://libagar.org/man3/AG_Fixed): Zoom operations.
- [**AG_StyleEditor**](https://libagar.org/man3/AG_StyleEditor): Edition of stylesheet rules.
- [**AG_StyleSheet**](https://libagar.org/man3/AG_StyleSheet): Gradients and border styles. New selector "Class > #id" for specific named instances of attached child widgets.
- [**AG_Surface**](https://libagar.org/man3/AG_Surface): Filters. Stencil operations.
- [**AG_Text**](https://libagar.org/man3/AG_Text): Provide a way to map core font names to known system fonts so porters have the option of adding dependencies to existing font packages in order to reduce the size of agar packages.
- [**AG_Textbox**](https://libagar.org/man3/AG_Textbox) & [**AG_Editable**](https://libagar.org/man3/AG_Editable): Extend SGR support. Syntax highlighting & rich-text editing methods.
- [**AG_WidgetPrimitives**](https://libagar.org/man3/AG_WidgetPrimitives): Dithering. Shadow effects.
- [**AG_Widget**](https://libagar.org/man3/AG_Widget): Provide a variation of the "zoom" feature to resize widgets individually.
- [**AG_Widget**](https://libagar.org/man3/AG_Widget): Threading optimizations. Cache the `VISIBLE` flag into an `int` so that culling / rendering and mouse-intersection tests can skip over invisible widgets without locking them.
- [**AG_Window**](https://libagar.org/man3/AG_Window): MRU API to simplify the process of remembering geometries. New gravity methods for autoplacing.
- [**cocoa**](https://libagar.org/man3/AG_DriverCocoa): Cursor operations. Clipboard integration.
- [**glx**](https://libagar.org/man3/AG_DriverGLX): Make drag-and-drop (`sockets` test) work correctly (possibly using `AG_WindowMove()`).
- [**wgl**](https://libagar.org/man3/AG_DriverWGL): Clipboard integration. Live resize.
- Fix [**Style Editor**](https://libagar.org/man3/AG_StyleEditor) and [**GUI Debugger**](https://libagar.org/man3/AG_GuiDebugger) crashes which would occur when attempting to iterate over widgets (in keyrepeat mode).
- Build system: Add [**CMake**](https://cmake.org) support.
- A software framebuffer driver for X Windows.

## [1.7.0] - 2022-10-
### Added
- [**sdl2fb**](https://libagar.org/man3/AG_DriverSDL2FB): New driver for SDL 2.0 (single-window; frame-buffer mode).
- [**sdl2gl**](https://libagar.org/man3/AG_DriverSDL2GL): New driver for SDL 2.0 (single-window; OpenGL mode).
- [**sdl2mw**](https://libagar.org/man3/AG_DriverSDL2MW): New driver for SDL 2.0 (multi-window; OpenGL mode).
- [**Single-window drivers**](https://libagar.org/man3/AG_DriverSw): New option `clampOnResize` (clamp active Agar windows against new display size).
- [**AG_Menu**](https://libagar.org/man3/AG_Menu): Implemented keyboard shortcuts. Added `AG_MENU_NO_BOOL_MSG` and `AG_MENU_FAST_BOOL_MSG` options.
- [**AG_Menu**](https://libagar.org/man3/AG_Menu): Improved the performance of the `AG_MenuView` renderer by caching state-dependent opaque labels (embedding any icons), so blending is no longer required. Improved appearance of `AG_MenuView`.
- [**AG_Menu**](https://libagar.org/man3/AG_Menu): When an `AG_MenuView` is detached, immediately invalidate the `view` pointers of its `AG_MenuItem`.
- [**AG_Radio**](https://libagar.org/man3/AG_Radio): New `HOMOGENOUS` option to divide space equally between items. Thanks Stephen!
- [**AG_Surface**](https://libagar.org/man3/AG_Surface): Embed 4 general-purpose 16-bit guides. Guides are useful for typography, graphics and other applications.
- [**AG_Text**](https://libagar.org/man3/AG_Text): The surface returned by `AG_TextRender()` now includes Ascent information under `Guide 0`.
- [**AG_Textbox**](https://libagar.org/man3/AG_Textbox) and [**AG_Editable**](https://libagar.org/man3/AG_Editable): Implement Undo/Redo functionality.
- [**AG_Tlist**](https://libagar.org/man3/AG_Tlist): Draw lines to express the relationship between tree items.
- [**AG_Tlist**](https://libagar.org/man3/AG_Tlist): New option `AG_TLIST_EXPAND_NODES`. Display nodes (items with children) initially in expanded state.
- [**AG_Tlist**](https://libagar.org/man3/AG_Tlist): New option `AG_TLIST_NO_SCALE_ICON`. Disable icon scaling when the icon width exceeds the standard width.
- [**AG_Tlist**](https://libagar.org/man3/AG_Tlist): New option `AG_TLIST_FIXED_HEIGHT`. Don't reset the item height on "font-changed".
- [**AG_Tlist**](https://libagar.org/man3/AG_Tlist): New item flag `AG_TLIST_ITEM_DISABLED`. Disable the item individually (prevent selection and display in "disabled" state).
- [**AG_Event**](https://libagar.org/man3/AG_Event): New macro [AG_EVENT_DUMP](https://libagar.org/man3/AG_EVENT_DUMP) to produce a listing of `argv[]` arguments on the console.
- [**AG_Widget**](https://libagar.org/man3/AG_Widget): New flag `AG_WIDGET_DISABLE_ON_ATTACH` and functions `AG_PushDisabledState()` and `AG_PopDisabledState`. Make attached child widgets start in DISABLED mode (without raising "widget-disabled").

sg/sg.c
/*
 * Copyright (c) 2005-2019 Julien Nadeau Carriere <vedge@csoft.net>
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE FOR
 * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
 * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
 * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
 * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
 * USE OF THIS SOFTWARE EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

/*
 * Agar-SG library initialization and implementation of the base
 * Scene Graph (SG) class.
 */

#include <agar/core/core.h>
#include <agar/sg/sg.h>
#include <agar/sg/sg_gui.h>

#include <math.h>
#include <string.h>

/* Standard Agar-SG classes */
void *sgStdClasses[] = {
	&sgClass,
	&sgNodeClass,
	&sgProgramClass,
/*	&sgCgProgramClass, <- Legacy */
	&sgTextureClass,
	&sgPaletteClass,
	&sgScriptClass,
	/* Non-geometrical */
	&sgDummyClass,
	&sgCameraClass,
	&sgLightClass,
	/* Reference geometry, editor controls */
	&sgGeomClass,
	&sgWidgetClass,
	&sgPointClass,
	&sgLineClass,
	&sgCircleClass,
	&sgSphereClass,
	&sgPlaneClass,
	&sgPolygonClass,
	&sgTriangleClass,
	&sgRectangleClass,
	/* BREP objects */
	&sgObjectClass,
	&sgPolyballClass,
	&sgPolyboxClass,
	/* Volumetric objects */
	&sgVoxelClass,
	/* Thin objects */
	&sgImageClass,
	NULL
};

/* File extension mappings */
const AG_FileExtMapping sgFileExtMap[] = {
	{ ".sg",	N_("Scene"),		&sgClass,		1 },
	{ ".sgs",	N_("Script"),		&sgScriptClass,		1 },
	{ ".sgt",	N_("Texture"),		&sgTextureClass,	1 },
	{ ".sgp",	N_("Palette"),		&sgPaletteClass,	1 },
	{ ".sgo",	N_("Object"),		&sgObjectClass,		0 },
};
const Uint sgFileExtCount = sizeof(sgFileExtMap) / sizeof(sgFileExtMap[0]);
int sgInitedSubsystem = 0;

/* Initialize the SG library. */
void
SG_InitSubsystem(void)
{
	void **cls;

	if (sgInitedSubsystem++ > 0)
		return;

	M_InitSubsystem();
	
	AG_RegisterNamespace("Agar-SG", "SG_", "https://libagar.org/");
	AG_RegisterFileExtMappings(sgFileExtMap, sgFileExtCount);

	for (cls = &sgStdClasses[0]; *cls != NULL; cls++)
		AG_RegisterClass(*cls);

	/* Initialize GUI facilities for SG_Edit() */
	SG_InitGUI();
}

/* Cleanup the SG library. */
void
SG_DestroySubsystem(void)
{
	void **cls;

	if (--sgInitedSubsystem > 0) {
		return;
	}
	SG_DestroyGUI();
	for (cls = &sgStdClasses[0]; *cls != NULL; cls++) {
		AG_UnregisterClass(*cls);
	}
	AG_UnregisterNamespace("Agar-SG");
}

/* Create a new SG object. */
SG *
SG_New(void *parent, const char *name, Uint flags)
{
	SG *sg;

	sg = Malloc(sizeof(SG));
	AG_ObjectInit(sg, &sgClass);
	if (name) {
		AG_ObjectSetNameS(sg, name);
	} else {
		OBJECT(sg)->flags |= AG_OBJECT_NAME_ONATTACH;
	}

	sg->root = (SG_Node *)SG_PointNew(sg, "_root", NULL);
	SG_PointSize(SGPOINT(sg->root), 0);
	sg->root->flags |= SG_NODE_HIDE;
	OBJECT(sg->root)->flags |= AG_OBJECT_INDESTRUCTIBLE;

	if (!(flags & SG_NO_DEFAULT_NODES)) {
		sg->def.cam = SG_CameraNew(sg->root, "Camera0");
		SG_Translate(sg->def.cam, 0.0, 0.0, 10.0);
		SG_CameraSetRotCtrlCircular(sg->def.cam, sg->root);
		OBJECT(sg->def.cam)->flags |= AG_OBJECT_INDESTRUCTIBLE;

		sg->def.lt[0] = SG_LightNew(sg->root, "Light0");
		SG_Translate(sg->def.lt[0], 30.0, 30.0, 30.0);
		OBJECT(sg->def.lt[0])->flags |= AG_OBJECT_INDESTRUCTIBLE;

		sg->def.lt[1] = SG_LightNew(sg->root, "Light1");
		SG_Translate(sg->def.lt[1], -30.0, -30.0, -30.0);
		OBJECT(sg->def.lt[1])->flags |= AG_OBJECT_INDESTRUCTIBLE;
	}
	
	AG_ObjectAttach(parent, sg);
	return (sg);
}

static void
Init(void *_Nonnull obj)
{
	SG *sg = obj;
	
	OBJECT(sg)->flags |= AG_OBJECT_REOPEN_ONLOAD;
	sg->flags = 0;
	sg->root = NULL;
	sg->def.cam = NULL;
	sg->def.lt[0] = NULL;
	sg->def.lt[1] = NULL;
	TAILQ_INIT(&sg->nodes);
}

static int
Save(void *_Nonnull obj, AG_DataSource *_Nonnull ds)
{
	SG *sg = obj;

	AG_WriteUint32(ds, sg->flags);
	return SG_NodeSave(sg, ds, sg->root);
}

static int
Load(void *_Nonnull obj, AG_DataSource *_Nonnull ds, const AG_Version *_Nonnull ver)
{
	SG *sg = obj;

	sg->flags = (Uint)AG_ReadUint32(ds);
	return SG_NodeLoad(sg, ds, NULL);
}

/* Search child nodes by name. */
SG_Node *
SG_SearchNodes(SG_Node *node, const char *name)
{
#ifdef AG_THREADS
	SG *sg = node->sg;
#endif
	SG_Node *chld, *rnode;

	AG_ObjectLock(sg);
	OBJECT_FOREACH_CLASS(chld, node, sg_node, "SG_Node:*") {
		if (strcmp(OBJECT(chld)->name, name) == 0) {
			AG_ObjectUnlock(sg);
			return (chld);
		} else {
			if ((rnode = SG_SearchNodes(chld, name)) != NULL) {
				AG_ObjectUnlock(sg);
				return (rnode);
			}
		}
	}
	AG_ObjectUnlock(sg);
	return (NULL);
}

/* Search entire graph for a node by name. */
void *
SG_FindNode(SG *sg, const char *name)
{
	void *rv;

	AG_ObjectLock(sg);
	if (strcmp(name, OBJECT(sg->root)->name) == 0) {
		rv = sg->root;
	} else {
		rv = (void *)SG_SearchNodes(sg->root, name);
	}
	AG_ObjectUnlock(sg);
	return (rv);
}

static void
ClearNodes(AG_Object *_Nonnull obj)
{
	AG_Object *cob, *ncob;
	
	AG_ObjectLock(obj);
	for (cob = TAILQ_FIRST(&obj->children);
	     cob != TAILQ_END(&obj->children);
	     cob = ncob) {
		ncob = TAILQ_NEXT(cob, cobjs);
		ClearNodes(cob);
	}
	AG_ObjectUnlock(obj);

	AG_ObjectDetach(obj);
	AG_ObjectDestroy(obj);
}

/*
 * Clear the scene, destroying all nodes (except the default nodes which
 * are reinitialized).
 */
void
SG_Clear(SG *sg)
{
	AG_Object *cob, *ncob;

	AG_ObjectLock(sg);
	for (cob = TAILQ_FIRST(&OBJECT(sg->root)->children);
	     cob != TAILQ_END(&OBJECT(sg->root)->children);
	     cob = ncob) {
		ncob = TAILQ_NEXT(cob, cobjs);
		if (cob == OBJECT(sg->def.cam) ||
		    cob == OBJECT(sg->def.lt[0]) ||
		    cob == OBJECT(sg->def.lt[1])) {
			continue;
		}
		ClearNodes(cob);
	}

	SG_Identity(sg->def.cam);
	SG_Translate(sg->def.cam, 0.0, 0.0, 10.0);
	SG_CameraSetRotCtrlCircular(sg->def.cam, sg->root);
	SG_Identity(sg->def.lt[0]);
	SG_Translate(sg->def.lt[0], 30.0, 30.0, 30.0);
	SG_Identity(sg->def.lt[1]);
	SG_Translate(sg->def.lt[1], -30.0, -30.0, -30.0);

	AG_ObjectUnlock(sg);
}

AG_ObjectClass sgClass = {
	"SG",
	sizeof(SG),
	{ 1,0 },
	Init,
	NULL,		/* reset */
	NULL,		/* destroy */
	Load,
	Save,
	SG_Edit
};
